### PR TITLE
Experiment with stronger typing for map

### PR DESCRIPTION
**Pull request by vladsud**
Is this interesting direction to explore?

It can take many forms, like
- stronger (whiteout undefined) typing if there is controlled creation path that populates initial values.
- users could mark some fields as readonly (likely useful only with above or having two alternative interfaces for same data).
- types baked into DDS (as an optional layer on top of it):
```typescript
class SharedMapTyped<T> extends SharedMap {
    readonly values: WithUndefined<T>;
}
```
- potentially alternative handle representation (if we find something interesting / useful here to reduce mental overload).

### CODE REVIEW

File: packages/dds/map/src/test/map.spec.ts [hunk 1]
```diff
@@ -32,6 +32,61 @@ function createLocalMap(id: string) {
...
 
+export interface ITest {
+    test: IFluidHandle<SharedMap>;
+    object: IFluidHandle<SharedMap>;
+    first: "second";
+    third: "fourth"
+    fifth: "sixth";
+}
+
+type WithUndefined<T> = {
```
Comment (anthony-murphy) [inline packages/dds/map/src/test/map.spec.ts:43]: the `Partial` utility class does the same thing

**Comment by anthony-murphy** [comment]
This is an interesting approach. i'd put thing in framework, not map itself, as it should be optional. another approach i'd thought about was something like 
```
TypedSharedMap<T>{

get(key: key of T): T[key];

set(key: keyof T, T[key];

}
```

**Comment by anthony-murphy** [comment]
@DLehenbauer curious about your thoughts here

**Comment by vladsud** [comment]
Either way will work fine I think. I prefer not to have .get/.set (stylistically), but that creates its own problems (you have to separate "values" object from map itself).
One thing to poke at though is semantics of this:

```javascript
map.get("a").foo = 1;
```

Is it clear to developers that it is supposed (not in current implementation - that's completely broken!) to be noop?
Do developers expect it to be noop? Or do they expect it to fail, and require some kind of explicit copy for this workflow to work:
```javascript
const value = map.get("a");
value.a = 1; // should this fail? be no-op?
const valuePrime = deepCopy(value); // or something else of same semantics
valuePrime.a = 1;
map.set("a", valuePrime);
```

The reason this question is important: it's a bit more obvious that value.a = 1 is noop (if we chose noop path here) with get() flow than with using my approach.

**Comment by vladsud** [comment]
Closing, but we should keep discussion going :)

**Pull request closed without merging**